### ISSUE TITLE
Unsure how to use libraries

### ISSUE THREAD

**Issue by JornVernee** [graalvm/sulong#673]
I have made a small example to test the linking of libraries as it is described in `README.md`:

> Libraries to load can be specified using the -l flag, as in a compiler:
   `mx su-run -lz test.bc`

I'm not sure in what format these libraries are supposed to be, I assumed that this is the `.su` files produced by `su-link`. I made the library with the commands:

    clang -c -emit-llvm -o dep.bc dep.c
    mx su-link -o dep.su dep.bc

I also tried the name `libdep.su` here.

Then I compile the main file with:

    clang -c -emit-llvm -o -test.bc test.c

and run with:

    mx su-run -ldep test.bc

Which throws a `RuntimeExcception` with the message: `could not resolve external symbol @func`. 

The library declares a single function:

    size_t func(const char* str);

That simply returns `strlen` of it's argument. I assumed that the call to `strlen` would fail, since the standard library would not be int bitcode. But then again, sulong might have recompiled it and included it internally. It was worth testing.

Here is the stacktrace:
```
 Exception in thread "main" java.lang.RuntimeException: LLVM error in assignment of %2 in first basic block in function @main in test.bc - could not resolve external symbol @func
	at com.oracle.truffle.llvm.nodes.base.LLVMBasicBlockNode.executeGetSuccessorIndex(LLVMBasicBlockNode.java:111)
	at com.oracle.truffle.llvm.nodes.control.LLVMDispatchBasicBlockNode.executeGeneric(LLVMDispatchBasicBlockNode.java:67)
	at com.oracle.truffle.llvm.nodes.func.LLVMFunctionStartNode.execute(LLVMFunctionStartNode.java:65)
	at com.oracle.truffle.api.impl.DefaultCallTarget.callDirectOrIndirect(DefaultCallTarget.java:63)
	at com.oracle.truffle.api.impl.DefaultDirectCallNode.call(DefaultDirectCallNode.java:43)
	at com.oracle.truffle.llvm.nodes.func.LLVMGlobalRootNode.executeIteration(LLVMGlobalRootNode.java:124)
	at com.oracle.truffle.llvm.nodes.func.LLVMGlobalRootNode.execute(LLVMGlobalRootNode.java:93)
	at com.oracle.truffle.api.impl.DefaultCallTarget.call(DefaultCallTarget.java:77)
	at com.oracle.truffle.llvm.nodes.base.LLVMMainFunctionReturnValueRootNode$LLVMMainFunctionReturnNumberRootNode.execute(LLVMMainFunctionReturnValueRootNode.java:82)
	at com.oracle.truffle.api.impl.DefaultCallTarget.callDirectOrIndirect(DefaultCallTarget.java:63)
	at com.oracle.truffle.api.impl.DefaultDirectCallNode.call(DefaultDirectCallNode.java:43)
	at com.oracle.truffle.api.vm.PolyglotRootNode$EvalRootNode.executeImpl(PolyglotRootNode.java:242)
	at com.oracle.truffle.api.vm.PolyglotRootNode.execute(PolyglotRootNode.java:68)
	at com.oracle.truffle.api.impl.DefaultCallTarget.call(DefaultCallTarget.java:77)
	at com.oracle.truffle.api.vm.PolyglotEngine.evalImpl(PolyglotEngine.java:637)
	at com.oracle.truffle.api.vm.PolyglotEngine.evalImpl(PolyglotEngine.java:617)
	at com.oracle.truffle.api.vm.PolyglotEngine.eval(PolyglotEngine.java:560)
	at com.oracle.truffle.llvm.LLVM.evaluateFromSource(LLVM.java:260)
	at com.oracle.truffle.llvm.LLVM.executeMain(LLVM.java:248)
	at com.oracle.truffle.llvm.LLVM.main(LLVM.java:235)
Caused by: java.lang.RuntimeException: could not resolve external symbol @func
	at com.oracle.truffle.llvm.runtime.NativeLookup.resolveAsNative(NativeLookup.java:267)
	at com.oracle.truffle.llvm.runtime.LLVMContext.resolveAsNativeFunction(LLVMContext.java:210)
	at com.oracle.truffle.llvm.nodes.func.LLVMDispatchNode.bindSymbol(LLVMDispatchNode.java:100)
	at com.oracle.truffle.llvm.nodes.func.LLVMDispatchNodeGen.executeAndSpecialize(LLVMDispatchNodeGen.java:143)
	at com.oracle.truffle.llvm.nodes.func.LLVMDispatchNodeGen.executeDispatch(LLVMDispatchNodeGen.java:74)
	at com.oracle.truffle.llvm.nodes.func.LLVMLookupDispatchNode.doDirect(LLVMLookupDispatchNode.java:68)
	at com.oracle.truffle.llvm.nodes.func.LLVMLookupDispatchNodeGen.executeAndSpecialize(LLVMLookupDispatchNodeGen.java:86)
	at com.oracle.truffle.llvm.nodes.func.LLVMLookupDispatchNodeGen.executeDispatch(LLVMLookupDispatchNodeGen.java:70)
	at com.oracle.truffle.llvm.nodes.func.LLVMCallNode.executeGeneric(LLVMCallNode.java:71)
	at com.oracle.truffle.llvm.nodes.api.LLVMExpressionNode.executeI64(LLVMExpressionNode.java:126)
	at com.oracle.truffle.llvm.nodes.func.LLVMCallUnboxNodeFactory$LLVMI64CallUnboxNodeGen.executeI64(LLVMCallUnboxNodeFactory.java:329)
	at com.oracle.truffle.llvm.nodes.vars.LLVMWriteNodeFactory$LLVMWriteI64NodeGen.executeGeneric(LLVMWriteNodeFactory.java:336)
	at com.oracle.truffle.llvm.nodes.base.LLVMBasicBlockNode.executeGetSuccessorIndex(LLVMBasicBlockNode.java:99)
	... 19 more
```

I noticed that this goes through something called `NativeLookup.resolveAsNative` so I also tried to use an `.so` file created with:

    mx su-clang -fPIC -shared -o dep.so dep.c

But the result is the same.

What format is the library supposed to be in?

**Comment by grimmerm** [comment]
We use `su-link` to create bitcode files/libraries that we can ship; it allows us to combine different files. That's probably not what you're looking for.

If you want to use a precompiled library from you C program running on Sulong, try:
`mx su-run -Dsulong.DynamicNativeLibraryPath=dep.so dep.c`

**Comment by JornVernee** [comment]
Thanks for the response, but that gives me the same error :/ I'm not sure, but this might be a naming issue. The function is name `func` (and has that name in the `.so` as well), but the error is trying to look up `@func`?

I'm also interested in knowing how this works. I thought sulong was just an llvm bitcode interpreter? But it seems there is a component that can access native code. I've traced the call to `ForeignAccess.sendRed(...)` which I don't have the source for (it that project Panama?). I assume that is also what is used to load the C standard library.

But how is the `-l` flag of `su-run`/`su-link` supposed to work? I can't find more about it in the repo's doc.

(sorry, I'm just curious)

**Comment by grimmerm** [comment]
Thanks for using Sulong ;-)
You are right, Sulong is an LLVM IR interpreter. 
You are also right, Sulong can indeed access native code (std library, or your library).
For these native calls, Sulong uses Truffle's native interface. Truffle's native interface allows you to load a library at runtime, lookup symbols, and then execute them.
That's what Sulong is doing in the NativeLookup class.

The naming seems correct, the @ is an llvm-internal thing that is prepended for each function.

Let's assume an example:
main.c (compiled to llvm ir using `clang -c -emit-llvm main.c -o main.bc`)

```
int foo();

int main() {
 return foo();
}
```

and a second file:
dep.c (compiled to a shared library, dep.so)

```
int foo() {
 return 42;
}
```

then you should be able to run this application using
`mx su-run -Dsulong.DynamicNativeLibraryPath=dep.so main.bc`

Hope this helps!

**Comment by JornVernee** [comment]
I started from scratch, and that  example worked. I had to give the full absolute path to `dep.so`. I was also able to run my code now. It seems I made another mistake somewhere. Thank you for your patience :)

Since I have the source for my library (`dep.so`), I guess it would be _better_ to first compile that into LLVM IR, then package `main.bc` and `dep.bc` with `su-link` (or perhaps `llvm-link`, which also worked), and then run the resulting `.su` (or `.bc`) file. Since the dependency's AST could also be optimized and inlined by the JIT compiler (taking advantage of speculative optimization)? (tbh, I'm not sure if `clang` emits the LLVM IR before optimization or after).
